396 - XForm validations don't fail if archive mode

mrsarm opened the pull request:
# Description

If the user is running medic-conf in `--archive` mode and the local API is not running, avoid failing when the XForm validations are requested, and instead warn with the following message:

    WARN The CHT Core instance is not running, no form will be checked before push

Also when trying to access other API endpoints and the API is down, there were messages logged like this: `ERROR Error trying to get couchdb config: RequestError: Error: Invalid URI "null//null@null/api/couch-config-attachments"` , now the message is:

    ERROR Error trying to get couchdb config: RequestError: Error: connect ECONNREFUSED 127.0.0.1:5988

medic/medic-conf#396

# Code review items

- Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- Tested: Unit and/or integration tests where appropriate
- Backwards compatible: Works with existing data and configuration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

garethbowen commented on src/lib/get-api-url.js:
```diff
@@ -13,17 +13,13 @@ const getApiUrl = (cmdArgs, env = {}) => {
 
   let instanceUrl;
-  if (cmdArgs.local) {
+  if (cmdArgs.local || cmdArgs.archive) {
```
I'm not 100% comfortable with this change, in part because it'll potentially affect other actions. In the case of form validation it may also validate against the wrong version of Enketo because it'll find whatever API you're running locally even if you didn't specifically request it and validate against that rather than the version you'll be pushing to.

I assume this is to fix the other actions mentioned in [your comment](https://github.com/medic/medic-conf/issues/396#issuecomment-823323674) but I think we should fix those by checking `environment.isArchiveMode` rather than returning an unexpected URL.

mrsarm replied:
Yes in archive mode medic-conf does connect with the local instance if it is available, but does not push any report at least so not sure why exactly it does the connection, but now with the new XForm validation the API URL is needed, unless there is another hack I can use.

I checked that the change does not affect the iteration as much I could understand it, eg. I changed the content of a report, then execute medic-conf in archive mode, and the change was not uploaded to the API as previous versions of medic-conf.

I can work in a different patch, but in the meantime @kennsippell could you please tell us whether our understanding is right and why the current implementation in master avoids the URL to be set in archive mode? I see you were one of the developer involved in this feature (https://github.com/medic/medic-conf/commit/5b8f8937fefe1389dfd1f9e6bb7aa6c9c6e1ca2d), that is why I'm asking for your advice please.

garethbowen reviewed (changes requested):
Nice one!

mrsarm commented:
The scope of this PR changed a bit an now is more a refactor of the `--archive` mode and some fixing around "environment" module that were causing the issues this PR tried to address.

First, some clarifications that were also added as comment in some of the files changed:

- Archive mode is only used by grunt in the task `build-config` at [Gruntfile.js#L579](https://github.com/medic/cht-core/blob/1c2108dc9179bc035149832e910d28ae248990bd/Gruntfile.js#L579), that is executed by default when pushing the first settings and the webapp assets to the CHT during development.
- I confirmed with Kenn that is not used by Apps developers in any other way.
- The task does not perform the upload, just the creation of the resources in a temporal folder that is pushed to the CouchDB later, so `--archive` mode shouldn't connect with the DB or the API in any way, but because a bug, it was doing that, and because a bug in the XForm validations, the validations were also trying to connect with the API in that mode.

So here are the list of changes in the PR along some clarifications:

- The archive module is implemented mocking the "api" and "PouchDB" modules, but some parts where not mocked and now they are (check archiving-api.js changes). That caused some calls to "escape the box", and medic-conf was actually calling the API when it was available, but only a readonly endpoint that tried to check what are the "compressible" types supported by Couch.
- Methods were added to the mock versions to emulate both the compressible types requests and the xform validations requests for the archive mode.
- In an earlier PR the deprecated method `url.parse(xxUrl)` was replaced by `new url.URL(xxUrl)`, but the replacement is not 100% compatible, and because in archive mode the apiUrl was mocked with an invalid '--archive mode' string, that caused the "Invalid URI --archive%20mode/api/..." errors seen in v3.5.0 and "Invalid URI null//null@n..." seen in earlier versions. Now the same URL used in `--local` mode is passed to the archive mode, tough no connection is performed because the fixes mentioned above.
- The environment module allowed to access getter values before initialized, returning default values that not necessarily matched the values passed to the CLI, now an exception is raised if that happens (but it shouldn't). This was potentially more danger that it looks like: because when JS evaluates the `undefined` value as `false` in a boolean expression, if the environment values were accessed earlier than initialized, the behavior could be different than the expected and without failing. It was happening in part of the implementation of the xform validations and in the tests where the environment module were not initialized nor mocked.

garethbowen reviewed (changes requested):
That's awesome! Very comprehensive.

If I understand correctly at least some of this issue occurred because the `archiving-api` didn't implement _all_ of the functions in `api`. If so, it'd be very easy to make that mistake again by adding something new to `api`. Can we add a unit test to compare the exported functions to make sure they're identical?

mrsarm commented:
> That's awesome! Very comprehensive.
> 
> If I understand correctly at least some of this issue occurred because the `archiving-api` didn't implement _all_ of the functions in `api`. If so, it'd be very easy to make that mistake again by adding something new to `api`. Can we add a unit test to compare the exported functions to make sure they're identical?

@garethbowen  not exactly, there is a fragment of code that prevent that stopping the execution of medic-conf, I just pushed a minor improvements over it: https://github.com/medic/medic-conf/pull/397/files#diff-f17bc036cb59910f33d92d2f663c324b1eed04c21c9e9ef6b5eef547c75688ffR178-R187 , with comments inline to make more clear the intention, and there was already tests to check whether archive flag makes the switch to the archive-api implementation, and I added others in this PR, like [this test](https://github.com/medic/medic-conf/pull/397/files#diff-72f368217e14f8978dd52b2d020d905d3892aecd29eabca3fa6759a4f9ab2a9bR55-R66).

However there were 2 similar problems among other fixed here:
1. There was a api call implemented outside the api module, therefore an implementation in the archive version wouldn't make any difference: https://github.com/medic/medic-conf/blob/ebdec0f1917654bb6c0ce5bbf42c1da28efb63dc/src/lib/warn-upload-overwrite.js#L27-L29
   I moved that `getCompressibleTypes` method to the api module and implemented the same in the archive mode with a mocked response (also added tests to cover the changes).
2. Environment values could be accessed before initialization, so an initialization of the api like this:
   ```js
   const api = require('./api')();
   // then once actions called...
   api.methodCall()...
   ```
   Instead of:
   ```js
   const api = require('./api');
   // then once actions initialized...
   api().methodCall()...    // notice the `()` that does the magic of switching between the different api implementations
   ```
   Caused the values from the environment module to be read before initialization in the api module code, and without failing but with an unexpected behavior: in this case, the real api implementation was used instead of the mocked one because `enviroment.isArchiveMode = undefined` instead of `true`, and in JS `undefined` is the same as `false` when evaluated as a boolean. I have removed the earlier initialization of the api module when importing with `require('./api')` ~`()`~, but if in a future implementation a bug like that is reintroduced, the access to the env variables not yet initialized would raise a clear error, and prevent the application to continue running because [this change](https://github.com/medic/medic-conf/pull/397/files#diff-8ae323656ca2f68e7894b791131d8311bbd74d6df219889ae79df180ea2a4e45R33-R41)), that is also covered by new [tests](https://github.com/medic/medic-conf/pull/397/files#diff-16b90bdf59597b2e054a73c223e2f716b9612323ca5baf9155769a3b937bdac7R10-R18).

garethbowen reviewed (approved):
Ok great. Thanks for that explanation! LGTM.
